src/pipeline.cpp

```diff
@@ -753,6 +755,47 @@ void NewPM::printTimers() {
     SI.getTimePasses().print();
 }
 
+void NewPM::registerCallbacks() {
+#ifdef ENABLE_TIMINGS
+    auto isMetaPass = [](StringRef PassID) {
+        auto pass_str = PassID.str();
+        if (pass_str.rfind("PassManager", 0) == 0) return true;
+        std::string end_str = "PassAdaptor";
+        if (pass_str.rfind(end_str) == pass_str.length() - end_str.length()) return true;
+        return false;
+    };
+
+    auto event = jl_timing_event_create("LLVM_JIT", "LLVM_OPT_PASS", "", __FILE__, __LINE__, 0);
+    PIC->registerBeforeNonSkippedPassCallback([event, isMetaPass](StringRef PassID, Any IR) {
+        if (isMetaPass(PassID)) return;
+        auto block = (jl_timing_block_t *)malloc(sizeof(jl_timing_block_t));
+        _jl_timing_block_init((char *)block, sizeof(jl_timing_block_t), event);
+        _jl_timing_block_start(block);
+        auto pass_str = PassID.str();
+        jl_timing_puts(block, pass_str.c_str());
+#ifdef USE_TRACY
+        TracyCZoneName(block->tracy_ctx, pass_str.c_str(), strlen(pass_str.c_str()));
+#endif
+    });
+    PIC->registerAfterPassCallback([isMetaPass](StringRef PassID, Any IR,
+                                          const PreservedAnalyses &PassPA) {
+        if (isMetaPass(PassID)) return;
+        if (jl_get_pgcstack() == NULL) return;
```

Comment:
Any opinions about this @topolarity

Reply:
We don't need the `jl_get_pgcstack() == NULL` check any more, but it'd be really nice to implement  https://github.com/JuliaLang/julia/pull/50926#pullrequestreview-1579096066 so that we're not just hiding LLVM work on its dedicated imaging threads

I'm going to take a whack at that later this week and see if we can land this with full thread visibility 👍

Reply:
Yeah you're absolutely right